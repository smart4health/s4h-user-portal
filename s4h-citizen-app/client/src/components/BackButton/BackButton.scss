.BackButton {
  &#{&} {
    width: 4rem;
    min-width: 4rem;
    height: 4rem;
    margin-right: 0.5em;
    color: var(--color-primary);

    svg {
      width: 100%;
      height: 100%;
    }
  }
}
