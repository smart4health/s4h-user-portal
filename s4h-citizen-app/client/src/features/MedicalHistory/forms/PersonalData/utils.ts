import { PersonalData } from '@d4l/s4h-fhir-xforms';
/* eslint-disable import/prefer-default-export */
import classNames from 'classnames';

export const valueSetUrls = {
  gender: 'http://hl7.org/fhir/ValueSet/administrative-gender',
  blood: 'http://loinc.org/vs/LL2419-1',
  rh: 'http://loinc.org/vs/LL360-9',
};

export type Option = {
  value: string;
  text: string;
};

const FEMALE = 'female';
const MALE = 'male';
const OTHER = 'other';
const UNKNOWN = 'unknown';

const BLOOD_0 = '0';
const BLOOD_A = 'A';
const BLOOD_B = 'B';
const BLOOD_AB = 'AB';

const BLOOD_RHESUS_FACTOR_POSITIVE = 'Positive';
const BLOOD_RHESUS_FACTOR_NEGATIVE = 'Negative';

const valueSetCodes: { [key: string]: string } = {
  'LA19708-9': BLOOD_0,
  'LA19709-7': BLOOD_B,
  'LA19710-5': BLOOD_A,
  'LA28449-9': BLOOD_AB,
  'LA6576-8': BLOOD_RHESUS_FACTOR_POSITIVE,
  'LA6577-6': BLOOD_RHESUS_FACTOR_NEGATIVE,
  female: FEMALE,
  male: MALE,
  other: OTHER,
  unknown: UNKNOWN,
};

export const mapValueSetCodeToTranslationKey = (code: string): string =>
  valueSetCodes[code];

export const fieldNames = {
  firstName: 'firstName',
  lastName: 'lastName',
  dateOfBirth: 'dateOfBirth',
  occupation: 'occupation',
  gender: 'gender',
  weight: 'weight',
  height: 'height',
  blood: 'blood',
  rh: 'rh',
};

export type FormValues = {
  firstName: string | undefined;
  lastName: string | undefined;
  dateOfBirth: string | undefined;
  occupation: string | undefined;
  gender: 'male' | 'female' | 'other' | 'unknown' | undefined;
  blood: 'LA19708-9' | 'LA19709-7' | 'LA19710-5' | 'LA28449-9' | undefined;
  rh: 'Positive' | 'Negative' | undefined;
  weight: string | undefined;
  height: string | undefined;
};

export type ValidationRules = {
  required?: boolean;
  min?: number;
  max?: number;
  maxlength?: number;
};

export const validationRules: { [key: string]: ValidationRules } = {
  [fieldNames.firstName]: {
    required: true,
  },
  [fieldNames.lastName]: {
    required: true,
  },
  [fieldNames.dateOfBirth]: {
    required: true,
  },
  [fieldNames.occupation]: {
    required: true,
  },
  [fieldNames.gender]: {
    required: true,
  },
  [fieldNames.weight]: {
    required: true,
    min: 1,
    max: 300,
  },
  [fieldNames.height]: {
    required: true,
    min: 1,
    max: 300,
  },
  [fieldNames.blood]: {
    required: true,
  },
  [fieldNames.rh]: {
    required: true,
  },
};

export const formToPersonalData = (formState: FormValues): PersonalData => {
  return {
    firstName: formState.firstName,
    lastName: formState.lastName,
    gender: formState.gender,
    dateOfBirth: formState.dateOfBirth,
    height: {
      value: parseInt(formState.height ? formState.height : '', 10),
      unit: 'cm',
    },
    weight: {
      value: parseInt(formState.weight ? formState.weight : '', 10),
      unit: 'kg',
    },
    bloodGroup: {
      system: 'http://loinc.org',
      code: formState.blood,
    },
    bloodRhesus: {
      system: 'http://loinc.org',
      code: formState.rh,
    },
    occupation: formState.occupation,
  };
};

export const personalDataToFormState = (personalData: PersonalData) => ({
  [fieldNames.firstName]: personalData?.firstName,
  [fieldNames.lastName]: personalData?.lastName,
  [fieldNames.dateOfBirth]: personalData?.dateOfBirth,
  [fieldNames.occupation]: personalData?.occupation,
  [fieldNames.gender]: personalData?.gender,
  [fieldNames.blood]: personalData?.bloodGroup?.code,
  [fieldNames.rh]: personalData?.bloodRhesus?.code,
  [fieldNames.weight]: personalData?.weight?.value,
  [fieldNames.height]: personalData?.height?.value,
});

export const getQuestionClass = (
  isReadOnly: boolean,
  hasBeenAnswered: boolean,
  additionalClasses = ''
) =>
  classNames(
    'MedicalHistoryQuestion',
    additionalClasses,
    {
      'MedicalHistoryQuestion--readonly': isReadOnly,
    },
    {
      'MedicalHistoryQuestion--answered': hasBeenAnswered,
    }
  );

export const formatDate = (year: string, month: string, day: string) =>
  `${year}${month ? `-${month}${day ? `-${day}` : ''}` : ''}`;
