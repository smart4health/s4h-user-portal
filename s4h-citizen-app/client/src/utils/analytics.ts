import ReactPiwik, { PiwikOptions } from 'react-piwik'; // '@types/react-piwik@latest' is not in the npm registry.
import config from '../config';
import settings, { SettingKeys } from './settings';

let matomo: any = null;

interface TrackingEvents {
  [key: string]: TrackingEvent;
}

type TrackingEvent = (string | number)[];

// The following mappings or names do not necessarily have to make sense.
// We're trying to map our events here to ones previously defined by mobile apps
// VAR_NAME_TERRA: ['Event category Matomo', 'Event action Matomo']
export const TRACKING_EVENTS: TrackingEvents = {
  DOCUMENT_UPLOAD_START: ['Documents - Save', 'Create document tapped'],
  DOCUMENT_UPLOAD_CANCEL: ['Documents - Save', 'Cancel create tapped'],
  DOCUMENT_UPLOAD_SUCCESS: ['Documents - Save', 'Document Save Success'],
  DOCUMENT_UPLOAD_ERROR: ['Documents - Save', 'Document Save Error'],
  DOCUMENT_EDIT_START: ['Documents - Editing', 'Edit document tapped'],
  DOCUMENT_EDIT_CANCEL: ['Documents - Editing', 'Cancel editing tapped'],
  DOCUMENT_EDIT_SUCCESS: ['Documents - Editing', 'Edit document success'],
  DOCUMENT_EDIT_ERROR: ['Documents - Editing', 'Edit document error'],
  DOCUMENT_DOWNLOAD_START: ['Documents - Download', 'Download document tapped'],
  DOCUMENT_DOWNLOAD_CANCEL: ['Documents - Download', 'Cancel download tapped'],
  DOCUMENT_DOWNLOAD_SUCCESS: ['Documents - Download', 'Download document success'],
  DOCUMENT_DOWNLOAD_ERROR: ['Documents - Download', 'Download document error'],
  DOCUMENT_ADD_START: ['Documents - Add', 'Add files tapped'],
  DOCUMENT_ADD_SUCCESS: ['Documents - Add', 'Add files success'],
  DOCUMENT_ADD_ERROR: ['Documents - Add', 'Add files error'],
  MEDICATION_DOWNLOAD_START: ['Medication - Download', 'Medication download tapped'],
  MEDICAL_HISTORY_PERSONAL_DATA_START: [
    'Medical History',
    'Edit personal_data tapped',
  ],
  MEDICAL_HISTORY_PERSONAL_DATA_ERROR: ['Medical History', 'personal_data error'],
  MEDICAL_HISTORY_PERSONAL_DATA_CANCEL: [
    'Medical History',
    'Cancel personal_data tapped',
  ],
  MEDICAL_HISTORY_PERSONAL_DATA_SUCCESS: [
    'Medical History',
    'Save personal_data tapped',
  ],
  MEDICAL_HISTORY_CONDITIONS_START: ['Medical History', 'Edit conditions tapped'],
  MEDICAL_HISTORY_CONDITIONS_ERROR: ['Medical History', 'conditions error'],
  MEDICAL_HISTORY_CONDITIONS_CANCEL: ['Medical History', 'Cancel conditions tapped'],
  MEDICAL_HISTORY_CONDITIONS_SUCCESS: ['Medical History', 'Save conditions tapped'],
  MEDICAL_HISTORY_RISK_FACTORS_START: [
    'Medical History',
    'Edit risk_factors tapped',
  ],
  MEDICAL_HISTORY_RISK_FACTORS_ERROR: ['Medical History', 'risk_factors error'],
  MEDICAL_HISTORY_RISK_FACTORS_CANCEL: [
    'Medical History',
    'Cancel risk_factors tapped',
  ],
  MEDICAL_HISTORY_RISK_FACTORS_SUCCESS: [
    'Medical History',
    'Save risk_factors tapped',
  ],
  MEDICAL_HISTORY_EMERGENCY_CONTACT_START: [
    'Medical History',
    'Edit emergency_contact tapped',
  ],
  MEDICAL_HISTORY_EMERGENCY_CONTACT_ERROR: [
    'Medical History',
    'emergency_contact error',
  ],
  MEDICAL_HISTORY_EMERGENCY_CONTACT_CANCEL: [
    'Medical History',
    'Cancel emergency_contact tapped',
  ],
  MEDICAL_HISTORY_EMERGENCY_CONTACT_SUCCESS: [
    'Medical History',
    'Save emergency_contact tapped',
  ],
  INITIATE_SHARING_START: ['Sharing', 'Start Sharing tapped'],
  INITIATE_SHARING_SUCCESS: ['Sharing', 'Start Sharing success'],
  SHARING_START: ['Sharing', 'Provide access tapped'],
  SHARING_SUCCESS: ['Sharing', 'Sharing success'],
  REVOKE_START: ['Sharing', 'Revoke tapped'],
  REVOKE_SUCCESS: ['Sharing', 'Revoke confirm tapped'],
  REVOKE_CANCEL: ['Sharing', 'Revoke cancel tapped'],
  NEW_SHARING_START: ['Sharing', 'Start new sharing session tapped'],
  OUTAPP_VIEWDATA_START: ['Sharing', 'View shared data tapped'],
  OUTAPP_VIEWDATA_SUCCESS: ['Sharing', 'View shared data success'],
  SHAREE_VIEW_DOWNLOAD_DATA: ['Sharee View', 'Download Data', 'Clicked'],
};

const isProduction: boolean = config.IS_PRODUCTION;
const isStaging: boolean = config.IS_STAGING;
const isLocalhost: boolean =
  config.ANALYTICS_ENABLE_LOCALHOST &&
  config.ANALYTICS_SITE_URL.startsWith('http://localhost:');

export const pushTrackingEvent = (trackingEvent: TrackingEvent) => {
  const isProduction = config.REACT_APP_ENVIRONMENT === 'production';

  if (!matomo || !ReactPiwik.push || !trackingEvent) {
    if (!isProduction) {
      console.log('TrackingEvent debug 🧐', trackingEvent);
    }
    return false;
  }
  ReactPiwik.push(['trackEvent', ...trackingEvent]);
  return true;
};
const initMatomo = () => {
  matomo = new ReactPiwik({
    url: config.ANALYTICS_SITE_URL,
    siteId: isProduction
      ? config.ANALYTICS_SITE_ID_PRODUCTION
      : isStaging
      ? config.ANALYTICS_SITE_ID_STAGING
      : config.ANALYTICS_SITE_ID_LOCALHOST,
    trackErrors: false,
    enableLinkTracking: true,
    jsFilename: 'matomo.js',
    phpFilename: 'matomo.php',
  } as PiwikOptions);

  ReactPiwik.push([
    'setCustomDimension',
    config.ANALYTICS_PRODUCTS_DIMENSION_KEY,
    config.ANALYTICS_PRODUCTS_DIMENSION_VALUE,
  ]);
  ReactPiwik.push([
    'setCustomDimension',
    config.ANALYTICS_PROJECTS_DIMENSION_KEY,
    config.ANALYTICS_PROJECTS_DIMENSION_VALUE,
  ]);
  ReactPiwik.push(['alwaysUseSendBeacon']);
  ReactPiwik.push(['enableCrossDomainLinking']);
  isProduction &&
    ReactPiwik.push(['setDomains', config.ANALYTICS_DOMAINS_PRODUCTION]);
  isStaging && ReactPiwik.push(['setDomains', config.ANALYTICS_DOMAINS_STAGING]);
  ReactPiwik.push(['trackPageView']);
  ReactPiwik.push(['enableHeartBeatTimer']);
};

const connectAnalytics = (browserHistory: any) => {
  if (!settings.acceptsTracking || (!isProduction && !isStaging && !isLocalhost)) {
    matomo && matomo.disconnectFromHistory();
    return false;
  }

  !matomo && initMatomo();
  matomo.connectToHistory(browserHistory);
  return true;
};

const setupAnalytics = (browserHistory: any) => {
  let isTracking = connectAnalytics(browserHistory);

  settings.onChange((key: string) => {
    if (key !== SettingKeys.ACCEPTS_TRACKING) {
      return;
    }

    const { acceptsTracking } = settings;
    if (isTracking !== acceptsTracking) {
      isTracking = connectAnalytics(browserHistory);
    }
  });

  return browserHistory;
};

export default setupAnalytics;
