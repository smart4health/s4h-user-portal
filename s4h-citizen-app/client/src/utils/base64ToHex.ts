export const base64ToHex = (str: string) => {
  for (
    var i = 0, bin = atob(str.replace(/[ \r\n]+$/, '')), hex = [];
    i < bin.length;
    ++i
  ) {
    let tmp = bin.charCodeAt(i).toString(16);
    if (tmp.length === 1) tmp = '0' + tmp;
    hex[hex.length] = tmp;
  }
  return hex.join('');
};
