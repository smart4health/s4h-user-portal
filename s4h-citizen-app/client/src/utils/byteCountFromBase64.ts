const byteCountFromBase64 = (base64: string): number =>
  base64.replace(/=/g, '').length * 0.75;
export default byteCountFromBase64;
