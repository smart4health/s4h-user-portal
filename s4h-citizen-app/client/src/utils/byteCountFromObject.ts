const byteCountFromObject = (data: object): number =>
  new Blob([JSON.stringify(data)]).size;

export default byteCountFromObject;
