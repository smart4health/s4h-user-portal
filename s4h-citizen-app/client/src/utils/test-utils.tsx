import { ThemeProvider as MuiThemeProvider } from '@material-ui/core/styles';
import { I18nextProvider } from 'react-i18next';
import { render } from '@testing-library/react';
import React, { ReactElement } from 'react';
import { configureStore } from '@reduxjs/toolkit';
import { Provider as ReduxProvider } from 'react-redux';
import { createMemoryHistory } from 'history';
import { Router } from 'react-router-dom';
// @ts-ignore
import { FeatureToggleProvider } from 'react-feature-toggles';
import rootReducer from '../redux/reducers';
import GHTheme from '../Material';
import { featureNames, FeatureNamesValues, Flags } from '../config/flags';
import i18n from '../mocks/i18n';
import { customizeInitialState } from '../store/state';
import { RootState } from '../types';
import { Provider as ReactWaterfallProvider } from '../store';

const featureFlagsAllEnabled: Flags = (
  Object.values(featureNames) as FeatureNamesValues[]
).reduce((accumulator, feature) => {
  accumulator[feature] = true;
  return accumulator;
}, {} as Flags);

/* Custom render function which helps us to have a configurable redux store with defined initial state*/

const history = createMemoryHistory();

const customRender = (
  ui: ReactElement,
  {
    // Ignoring below as a way to type this rightly to Partial<StoreState> was not found
    // @ts-ignore
    initialState,
    store = configureStore({
      reducer: rootReducer,
      preloadedState: initialState,
      middleware: getDefaultMiddleware => getDefaultMiddleware(),
    }),
    extendedWaterfallState = {} as Partial<RootState>,
    ...renderOptions
  } = {}
) => {
  const AllTheProviders: React.FC = props => {
    customizeInitialState({
      flags: featureFlagsAllEnabled,
      ...extendedWaterfallState,
    });
    return (
      <Router history={history}>
        <FeatureToggleProvider featureToggleList={featureFlagsAllEnabled}>
          <ReduxProvider store={store}>
            <ReactWaterfallProvider>
              <I18nextProvider i18n={i18n}>
                <MuiThemeProvider theme={GHTheme}>{props.children}</MuiThemeProvider>
              </I18nextProvider>
            </ReactWaterfallProvider>
          </ReduxProvider>
        </FeatureToggleProvider>
      </Router>
    );
  };

  return render(ui, { wrapper: AllTheProviders, ...renderOptions });
};

export * from '@testing-library/react';
export { customRender as render, history };
