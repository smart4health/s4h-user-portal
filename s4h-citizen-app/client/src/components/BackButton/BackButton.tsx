import Button from '@material-ui/core/Button';
import ArrowBack from '@material-ui/icons/ArrowBack';
import React, { MouseEventHandler } from 'react';
import { Link } from 'react-router-dom';
import './BackButton.scss';

type Props = {
  toRoute?: string;
  onClick?: MouseEventHandler;
};

const BackButton = ({ toRoute, onClick }: Props) => {
  if (toRoute) {
    return (
      <Button
        className="Button BackButton"
        variant="text"
        component={Link}
        onClick={onClick}
        to={toRoute}
      >
        <ArrowBack />
      </Button>
    );
  }
  return (
    <Button className="Button BackButton" variant="text" onClick={onClick}>
      <ArrowBack />
    </Button>
  );
};

export default BackButton;
