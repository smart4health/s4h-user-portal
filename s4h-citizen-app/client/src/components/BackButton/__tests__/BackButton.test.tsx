import React from 'react';
import { render } from '../../../utils/test-utils';
import BackButton from '../BackButton';

describe('BackButton ', () => {
  it('renders correctly', () => {
    const { container } = render(<BackButton />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
