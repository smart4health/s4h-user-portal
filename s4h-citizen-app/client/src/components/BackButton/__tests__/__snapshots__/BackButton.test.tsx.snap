// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BackButton  renders correctly 1`] = `
<button
  class="MuiButtonBase-root MuiButton-root MuiButton-text Button BackButton"
  tabindex="0"
  type="button"
>
  <span
    class="MuiButton-label"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
      />
    </svg>
  </span>
</button>
`;
